enterprise/backend/src/metabase_enterprise/workspaces/models/workspace.clj

```diff
@@ -50,24 +51,38 @@
     (mi/instances-with-hydrated-data wses k (ws-contents ids) :id {:default {}})))
 
 (defn archive!
-  "Archive a workspace. Destroys database isolation resources and sets archived_at."
+  "Archive a workspace. Destroys database isolation resources, revokes access grants, and sets archived_at."
   [workspace]
-  (let [database (t2/select-one :model/Database (:database_id workspace))]
+  (let [database     (t2/select-one :model/Database (:database_id workspace))
+        workspace-id (:id workspace)]
     (ws.isolation/destroy-workspace-isolation! database workspace)
-    (t2/update! :model/Workspace (:id workspace) {:archived_at [:now]})))
+    ;; Mark all inputs as ungranted since the user was dropped
+    (t2/update! :model/WorkspaceInput {:workspace_id workspace-id}
+                {:access_granted false})
+    (t2/update! :model/Workspace workspace-id {:archived_at [:now]})))
 
 (defn unarchive!
-  "Unarchive a workspace. Re-initializes database isolation resources and clears archived_at."
+  "Unarchive a workspace. Re-initializes database isolation resources, re-grants read access
+   to source tables, and clears archived_at."
   [workspace]
   (let [database         (t2/select-one :model/Database (:database_id workspace))
-        isolation-result (ws.isolation/ensure-database-isolation! workspace database)]
-    (t2/update! :model/Workspace (:id workspace)
-                (merge {:archived_at nil}
-                       (select-keys isolation-result [:schema :database_details])))))
+        isolation-result (ws.isolation/ensure-database-isolation! workspace database)
+        workspace-id     (:id workspace)
+        _                (t2/update! :model/Workspace workspace-id
+                                     (merge {:archived_at nil}
+                                            (select-keys isolation-result [:schema :database_details])))
+        ;; Re-fetch workspace to get updated database_details and schema for granting permissions
+        updated-ws       (t2/select-one :model/Workspace workspace-id)]
+    ;; Re-grant read access to source tables (inputs were marked ungranted during archive)
+    (ws.impl/sync-grant-accesses! updated-ws)))
 
 (defn delete!
   "Delete a workspace. Destroys database isolation resources and deletes the workspace."
   [workspace]
   (let [database (t2/select-one :model/Database (:database_id workspace))]
+    ;; (Ngoc - 2025-12-17) not sure if this is necessary here because you have to archive in order to delete, thus isolation has been
+    ;; dropped already
     (ws.isolation/destroy-workspace-isolation! database workspace)
+    (t2/update! :model/WorkspaceInput {:workspace_id (:id workspace)}
+                {:access_granted false})
```

Comment:
Note that we could fall out of sync here if there is a partial failure in the destroy method.

I _think_ this is the right way around - safer to overestimate our access than have access we don't know about. Then again, it would be simpler to have underestimated them when we want to idempotently grant what's missing.

Something to ponder, but no action required.